pkg/prometheus/agent/statefulset.go

```diff
@@ -0,0 +1,483 @@
+			operator.WatchedDirectories(watchedDirectories), operator.VolumeMounts(configReloaderVolumeMounts),
+			operator.Shard(shard),
+			operator.ImagePullPolicy(cpf.ImagePullPolicy),
+		),
+	}, additionalContainers...)
+
+	containers, err := k8sutil.MergePatchContainers(operatorContainers, cpf.Containers)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to merge containers spec")
+	}
+
+	// PodManagementPolicy is set to Parallel to mitigate issues in kubernetes: https://github.com/kubernetes/kubernetes/issues/60164
+	// This is also mentioned as one of limitations of StatefulSets: https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#limitations
+	return &appsv1.StatefulSetSpec{
+		ServiceName:         governingServiceName,
+		Replicas:            cpf.Replicas,
+		PodManagementPolicy: appsv1.ParallelPodManagement,
+		UpdateStrategy: appsv1.StatefulSetUpdateStrategy{
+			Type: appsv1.RollingUpdateStatefulSetStrategyType,
+		},
+		MinReadySeconds: minReadySeconds,
+		Selector: &metav1.LabelSelector{
+			MatchLabels: finalSelectorLabels,
+		},
+		Template: v1.PodTemplateSpec{
+			ObjectMeta: metav1.ObjectMeta{
+				Labels:      finalLabels,
+				Annotations: podAnnotations,
+			},
+			Spec: v1.PodSpec{
+				Containers:                    containers,
+				InitContainers:                initContainers,
+				SecurityContext:               cpf.SecurityContext,
+				ServiceAccountName:            cpf.ServiceAccountName,
+				AutomountServiceAccountToken:  &boolTrue,
+				NodeSelector:                  cpf.NodeSelector,
+				PriorityClassName:             cpf.PriorityClassName,
+				TerminationGracePeriodSeconds: &terminationGracePeriod,
+				Volumes:                       volumes,
+				Tolerations:                   cpf.Tolerations,
+				Affinity:                      cpf.Affinity,
+				TopologySpreadConstraints:     cpf.TopologySpreadConstraints,
+				HostAliases:                   operator.MakeHostAliases(cpf.HostAliases),
+				HostNetwork:                   cpf.HostNetwork,
+			},
+		},
+	}, nil
+}
+
+func makeStatefulSetService(p *monitoringv1alpha1.PrometheusAgent, config operator.Config) *v1.Service {
+	p = p.DeepCopy()
+
+	if p.Spec.PortName == "" {
+		p.Spec.PortName = prompkg.DefaultPortName
+	}
+
+	svc := &v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: governingServiceName,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					Name:       p.GetName(),
+					Kind:       p.Kind,
+					APIVersion: p.APIVersion,
+					UID:        p.GetUID(),
+				},
+			},
+			Labels: config.Labels.Merge(map[string]string{
+				"operated-prometheus": "true",
+			}),
+		},
+		Spec: v1.ServiceSpec{
+			ClusterIP: "None",
+			Ports: []v1.ServicePort{
+				{
+					Name:       p.Spec.PortName,
+					Port:       9090,
+					TargetPort: intstr.FromString(p.Spec.PortName),
+				},
+			},
+			Selector: map[string]string{
+				"app.kubernetes.io/name": "prometheus-agent",
```

Comment:
If we have a namespace where we deploy more than one Prometheus Agent CR isn't this going to send traffic to all instances? Shouldn't we include the name of the CR (`"app.kubernetes.io/instance"`) as we do for the statefulsets?

Reply:
Yes in principle though this service is used as the statefulset's [governing service](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#stable-network-id) which exists for creating DNS entries rather than load-balancing.
It takes the same [approach](https://github.com/prometheus-operator/prometheus-operator/blob/32ebc58aeaf6266946d67c4b446f7a36265e543d/pkg/prometheus/statefulset.go#L123-L169) than the Prometheus CRD controller but we might be able to do better: like you said, 2 objects in the same namespace would be aggregated in the same `*-operated` service...